File: qwen2.5-coder/ordenTopologico/orden_topologico.js
class Graph {
    constructor(vertices) {
        this.V = vertices; // Número de vértices
        this.adjList = new Map(); // Lista adyacente del grafo

        for (let i = 0; i < vertices; i++) {
            this.adjList.set(i, []);
        }
    }

    // Método para agregar una arista dirigida al grafo
    addEdge(v, w) {
        this.adjList.get(v).push(w);
    }

    // Método para encontrar el orden topológico del grafo usando Kahn's Algorithm
    topologicalSort() {
        const inDegree = Array(this.V).fill(0); // Grados entrantes de cada vértice

        // Calcular los grados entrantes de cada vértice
        this.adjList.forEach((neighbors) => {
            neighbors.forEach((neighbor) => {
                inDegree[neighbor]++;
            });
        });

        const queue = []; // Cola para almacenar vértices con grado entrante 0

        // Agregar todos los vértices con grado entrante 0 a la cola
        for (let i = 0; i < this.V; i++) {
            if (inDegree[i] === 0) {
                queue.push(i);
            }
        }

        const topologicalOrder = []; // Almacenar el orden topológico

        while (queue.length > 0) {
            const u = queue.shift(); // Extraer un vértice de la cola
            topologicalOrder.push(u);

            // Reducir los grados entrantes de los vecinos del vértice extraído
            this.adjList.get(u).forEach((neighbor) => {
                inDegree[neighbor]--;

                // Si el grado entrante del vecino se vuelve 0, agregarlo a la cola
                if (inDegree[neighbor] === 0) {
                    queue.push(neighbor);
                }
            });
        }

        // Verificar si el grafo es acíclico
        if (topologicalOrder.length !== this.V) {
            console.log("El grafo contiene un ciclo, no se puede realizar orden topológico.");
            return [];
        }

        return topologicalOrder;
    }
}

// Ejemplo de uso
const g = new Graph(6);
g.addEdge(5, 2);
g.addEdge(5, 0);
g.addEdge(4, 0);
g.addEdge(4, 1);
g.addEdge(2, 3);
g.addEdge(3, 1);

console.log("Orden topológico del grafo:");
console.log(g.topologicalSort());

File: qwen2.5-coder/ordenTopologico/orden_topologico.rs
use std::collections::{VecDeque, HashMap};

// Representación de un nodo y su lista de adyacencias.
struct Node {
    id: usize,
    in_degree: u32, // Grado entrante del nodo.
    neighbors: Vec<usize>,
}

impl Node {
    fn new(id: usize) -> Self {
        Node { id, in_degree: 0, neighbors: Vec::new() }
    }

    fn add_neighbor(&mut self, neighbor_id: usize) {
        self.neighbors.push(neighbor_id);
    }
}

// Estructura para representar el grafo.
pub struct Graph {
    nodes: HashMap<usize, Node>,
}

impl Graph {
    pub fn new() -> Self {
        Graph { nodes: HashMap::new() }
    }

    // Añade un nodo al grafo con el identificador dado.
    pub fn add_node(&mut self, id: usize) {
        if !self.nodes.contains_key(&id) {
            self.nodes.insert(id, Node::new(id));
        }
    }

    // Añade una arista desde 'from' hasta 'to'.
    pub fn add_edge(&mut self, from: usize, to: usize) {
        let from_node = self.nodes.get_mut(&from).unwrap();
        let to_node = self.nodes.get_mut(&to).unwrap();

        from_node.add_neighbor(to);
        to_node.in_degree += 1;
    }

    // Realiza el recorrido topológico utilizando el algoritmo de Kahn.
    pub fn kahn_topo_sort(&self) -> Option<Vec<usize>> {
        let mut sorted = Vec::new();
        let mut queue = VecDeque::new();

        // Encuentra todos los nodos con grado entrante 0 y añade a la cola.
        for (id, node) in &self.nodes {
            if node.in_degree == 0 {
                queue.push_back(*id);
            }
        }

        while let Some(node_id) = queue.pop_front() {
            sorted.push(node_id);

            // Reduce el grado entrante de los vecinos del nodo actual.
            if let Some(node) = self.nodes.get(&node_id) {
                for &neighbor in &node.neighbors {
                    if let Some(neighbor_node) = self.nodes.get_mut(&neighbor) {
                        neighbor_node.in_degree -= 1;
                        if neighbor_node.in_degree == 0 {
                            queue.push_back(neighbor);
                        }
                    }
                }
            }
        }

        // Verifica si el grafo tiene un orden topológico válido.
        if sorted.len() != self.nodes.len() {
            None
        } else {
            Some(sorted)
        }
    }
}

// Ejemplo de uso:
fn main() {
    let mut graph = Graph::new();
    graph.add_node(1);
    graph.add_node(2);
    graph.add_node(3);
    graph.add_node(4);

    graph.add_edge(1, 2);
    graph.add_edge(1, 3);
    graph.add_edge(2, 4);
    graph.add_edge(3, 4);

    if let Some(topo_order) = graph.kahn_topo_sort() {
        println!("Orden topológico: {:?}", topo_order);
    } else {
        println!("El grafo no tiene un orden topológico válido.");
    }
}